src/x/instrument/config_prometheus.go

```diff
@@ -0,0 +1,323 @@
+	// objectives to be used by the reporter.
+	DefaultSummaryObjectives []prometheus.SummaryObjective `yaml:"defaultSummaryObjectives"`
+
+	// OnError specifies what to do when an error either with listening
+	// on the specified listen address or registering a metric with the
+	// Prometheus. By default the registerer will panic.
+	OnError string `yaml:"onError"`
+}
+
+// HistogramObjective is a Prometheus histogram bucket.
+// See: https://godoc.org/github.com/prometheus/client_golang/prometheus#HistogramOpts
+type HistogramObjective struct {
+	Upper float64 `yaml:"upper"`
+}
+
+// SummaryObjective is a Prometheus summary objective.
+// See: https://godoc.org/github.com/prometheus/client_golang/prometheus#SummaryOpts
+type SummaryObjective struct {
+	Percentile   float64 `yaml:"percentile"`
+	AllowedError float64 `yaml:"allowedError"`
+}
+
+// PrometheusConfigurationOptions allows some programatic options, such as using a
+// specific registry and what error callback to register.
+type PrometheusConfigurationOptions struct {
+	// Registry if not nil will specify the specific registry to use
+	// for registering metrics.
+	Registry *prom.Registry
+	// ExternalRegistries if set (with combination of a specified Registry)
+	// will also
+	ExternalRegistries []PrometheusExternalRegistry
+	// HandlerOpts is the reporter HTTP handler options, not specifying will
+	// use defaults.
+	HandlerOpts promhttp.HandlerOpts
+	// OnError allows for customization of what to do when a metric
+	// registration error fails, the default is to panic.
+	OnError func(e error)
+}
+
+// PrometheusExternalRegistry is an external Prometheus registry
+// to also expose as part of the handler.
+type PrometheusExternalRegistry struct {
+	// Registry is the external prometheus registry to list.
+	Registry *extprom.Registry
+	// SubScope will add a prefix to all metric names exported by
+	// this registry.
+	SubScope string
+}
+
+// NewReporter creates a new M3 Prometheus reporter from this configuration.
+func (c PrometheusConfiguration) NewReporter(
+	configOpts PrometheusConfigurationOptions,
+) (prometheus.Reporter, error) {
+	var opts prometheus.Options
+
+	if configOpts.Registry != nil {
+		opts.Registerer = configOpts.Registry
+	}
+
+	if configOpts.OnError != nil {
+		opts.OnRegisterError = configOpts.OnError
+	} else {
+		switch c.OnError {
+		case "stderr":
+			opts.OnRegisterError = func(err error) {
+				fmt.Fprintf(os.Stderr, "tally prometheus reporter error: %v\n", err)
+			}
+		case "log":
+			opts.OnRegisterError = func(err error) {
+				log.Printf("tally prometheus reporter error: %v\n", err)
+			}
+		case "none":
+			opts.OnRegisterError = func(err error) {}
+		default:
+			opts.OnRegisterError = func(err error) {
+				panic(err)
+			}
+		}
+	}
+
+	switch c.TimerType {
+	case "summary":
+		opts.DefaultTimerType = prometheus.SummaryTimerType
+	case "histogram":
+		opts.DefaultTimerType = prometheus.HistogramTimerType
+	}
```

Comment:
Should this have a default case?

Reply:
The default value for `DefaultTimerType` is actually valid hence why it is not handled here.